### PR TITLE
New utilities

### PR DESCRIPTION
**Pull request by mattjones315**
New utilities that will be useful for real and simulated pipelines -- 

(1) Bootstrapping functionality:
-- given a character matrix, generate several bootstrap replicates assuming independent characters. 
-- given an alleletable, generate bootstrap character matrices respecting the intBCs (i.e. sample intBCs rather than individual cut sites)

(2) Various functions for converting between data structures intrinsic to our lineage tracing technology:
-- alleletable to character matrix (with integer representations)
-- alleletable to lineage profile (this is like a character matrix, but with actual indel identities represented)
-- lineage profile to character matrix

(3) Utility for generating empirical indel probabilities from alleletables.

Changed code is distributed between `cassiopeia.data.utilities.py` and `cassiopeia.preprocess.utilities.py`. Tests are represented in the `preprocess_tests` and `data_tests` directories.

### CODE REVIEW

File: cassiopeia/data/utilities.py [hunk 1]
```diff
@@ -165,4 +174,131 @@ def _compute_dissimilarity_map(cm, C, missing_state_indicator, nb_weights):
...
-    return _compute_dissimilarity_map(cm, C, missing_state_indicator, nb_weights)
+    return _compute_dissimilarity_map(
+        cm, C, missing_state_indicator, nb_weights
+    )
+
+
+def sample_bootstrap_character_matrices(
+    character_matrix: pd.DataFrame,
+    prior_probabilities: Optional[Dict[int, Dict[int, float]]] = None,
+    B: int = 10,
```
Comment (sprillo) [inline cassiopeia/data/utilities.py:185]: I'd prefer a lowercase and more expressive name than `B`, e.g. `n_bootstraps` or `num_bootstraps` etc.

File: cassiopeia/data/utilities.py [hunk 2]
```diff
@@ -165,4 +174,131 @@ def _compute_dissimilarity_map(cm, C, missing_state_indicator, nb_weights):
...
+    random_state: Optional[np.random.RandomState] = None,
+) -> List[Tuple[pd.DataFrame, Dict[int, Dict[int, float]]]]:
+    """Generates bootstrapped character matrices from a character matrix.
+
+    Ingests a character matrix and randomly creates bootstrap samples by
+    sampling with replacement characters. Each bootstrapped character matrix,
```
Comment (sprillo) [inline cassiopeia/data/utilities.py:191]: `sampling with replacement characters` => This sounds better: `sampling characters with replacement`

File: cassiopeia/data/utilities.py [hunk 3]
```diff
@@ -165,4 +174,131 @@ def _compute_dissimilarity_map(cm, C, missing_state_indicator, nb_weights):
...
+        cm, C, missing_state_indicator, nb_weights
+    )
+
+
+def sample_bootstrap_character_matrices(
+    character_matrix: pd.DataFrame,
+    prior_probabilities: Optional[Dict[int, Dict[int, float]]] = None,
+    B: int = 10,
+    random_state: Optional[np.random.RandomState] = None,
+) -> List[Tuple[pd.DataFrame, Dict[int, Dict[int, float]]]]:
```
Comment (sprillo) [inline cassiopeia/data/utilities.py:187]: `new_priors` can be `None`, so the return type should be `List[Tuple[pd.DataFrame, Optional[Dict[int, Dict[int, float]]]]]`.

File: cassiopeia/data/utilities.py [hunk 4]
```diff
@@ -165,4 +174,131 @@ def _compute_dissimilarity_map(cm, C, missing_state_indicator, nb_weights):
...
+
+        bootstrap_samples.append((bootstrapped_character_matrix, new_priors))
+
+    return bootstrap_samples
+
+
+def sample_bootstrap_allele_tables(
+    allele_table: pd.DataFrame,
+    indel_priors: Optional[pd.DataFrame] = None,
+    B: int = 10,
```
Comment (sprillo) [inline cassiopeia/data/utilities.py:238]: Same comment as before.

File: cassiopeia/data/utilities.py [hunk 5]
```diff
@@ -165,4 +174,131 @@ def _compute_dissimilarity_map(cm, C, missing_state_indicator, nb_weights):
...
+    random_state: Optional[np.random.RandomState] = None,
+):
```
Comment (sprillo) [inline cassiopeia/data/utilities.py:240]: Missing return type here.

File: cassiopeia/preprocess/utilities.py [hunk 6]
```diff
@@ -4,16 +4,18 @@
...
+from typing import Dict, List, Optional, Tuple
+
 
 from collections import defaultdict, OrderedDict
 import Levenshtein
 import matplotlib
 import matplotlib.pyplot as plt
 import numpy as np
 import pandas as pd
+import pylab
```
Comment (sprillo) [inline cassiopeia/preprocess/utilities.py:16]: Should this be added to `setup.py` / `requirements.txt`?

File: cassiopeia/preprocess/utilities.py [hunk 7]
```diff
@@ -387,13 +389,17 @@ def convert_alleletable_to_character_matrix(
     for sample in alleletable.index:
         cell = alleletable.loc[sample, "cellBC"]
         intBC = alleletable.loc[sample, "intBC"]
-        cut_sites = ["_r1", "_r2", "_r3"]
+        cut_sites = [
+            f"_{column}"
```
Comment (sprillo) [inline cassiopeia/preprocess/utilities.py:393]: is this some standard format? Or is the alleletable format custom?

File: cassiopeia/preprocess/utilities.py [hunk 8]
```diff
@@ -451,36 +458,242 @@ def convert_alleletable_to_character_matrix(
...
+        allele_table: AlleleTable.
+
+    Returns:
+        An NxM lineage profile.
+    """
+
+    cutsites = [
+        column
+        for column in allele_table.columns
+        if bool(re.search(r"r\d", column))
```
Comment (sprillo) [inline cassiopeia/preprocess/utilities.py:513]: This logic seems to be duplicated from before, maybe we should factor it out into a function `def cutsites(...)`.

File: cassiopeia/preprocess/utilities.py [hunk 9]
```diff
@@ -451,36 +458,242 @@ def convert_alleletable_to_character_matrix(
...
+    agg_recipe = dict(
+        zip([cutsite for cutsite in cutsites], ["unique"] * len(cutsites))
+    )
+    g = allele_table.groupby(["cellBC", "intBC"]).agg(agg_recipe)
+    intbcs = allele_table["intBC"].unique()
+
+    # create mutltindex df by hand
+    i1 = []
+    for i in intbcs:
+        i1 += [i] * 3
```
Comment (sprillo) [inline cassiopeia/preprocess/utilities.py:524]: Any reason for this hardcoded `3`? I'd prefer to give this a name and place it in some `globals.py` file. (We should ideally also do this for the default missing character state in our functions, which is currently hardcoded to `-1` everywhere. This helps a lot when e.g. searching through code. Plus, hardcoded constants are a source of nasty bugs.)

File: cassiopeia/preprocess/utilities.py [hunk 10]
```diff
@@ -451,36 +458,242 @@ def convert_alleletable_to_character_matrix(
...
+        allele_table: Alleletable
+        grouping_variables: Variables to stratify data by, to treat as
+            independent groups in counting indel occurrences. These must be
+            columns in the allele table
+
+    Returns:
+        A DataFrame mapping indel identities to the probability.
+    """
+
+    cut_sites = [
```
Comment (sprillo) [inline cassiopeia/preprocess/utilities.py:669]: Triplicated.

File: test/data_tests/utilities_test.py [hunk 11]
```diff
@@ -0,0 +1,218 @@
...
+                "TTT": 0.05,
+                "ATA": 0.1,
+            },
+            orient="index",
+            columns=["freq"],
+        )
+
+    def test_bootstrap_character_matrices_no_priors(self):
+
+        random_state = np.random.RandomState(123431235)
```
Comment (sprillo) [inline test/data_tests/utilities_test.py:72]: I like your style.

**Review by sprillo — APPROVED** [review]
Some minor comments.

**Review by richardyz98 — APPROVED** [review]

File: test/data_tests/utilities_test.py (hunk 11 above)
Comment (richardyz98) [reply test/data_tests/utilities_test.py:78]: Is this an inside joke I've been thinking about this all afternoon

**Pull request merged**